pandas/core/groupby.py

```diff
@@ -3055,7 +3055,7 @@ def transform(self, func, *args, **kwargs):
         # we have a numeric dtype, as these are *always* udfs
         # the cython take a different path (and casting)
         dtype = self._selected_obj.dtype
-        if is_numeric_dtype(dtype):
+        if is_numeric_dtype(dtype) and not is_bool_dtype(result.dtype):
             result = maybe_downcast_to_dtype(result, dtype)
```

Comment:
instead make this change *inside* ``maybe_downcast_to_dtype`` , roughly around https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/cast.py#L113